File: redboost_backend/src/main/java/team/project/redboost/services/TypeFormationService.java
package team.project.redboost.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import team.project.redboost.dto.CreateTypeFormationRequest;
import team.project.redboost.dto.TypeFormationResponse;
import team.project.redboost.entities.TypeFormation;
import team.project.redboost.repositories.TypeFormationRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class TypeFormationService {

    private final TypeFormationRepository typeFormationRepository;

    /**
     * Get all type formations
     */
    public List<TypeFormationResponse> getAllTypes() {
        return typeFormationRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Get a specific type formation by ID
     */
    public TypeFormationResponse getTypeById(Long id) {
        TypeFormation type = typeFormationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Type de formation not found with id: " + id));
        return mapToResponse(type);
    }

    /**
     * Create a new type formation
     */
    @Transactional
    public TypeFormationResponse createType(CreateTypeFormationRequest request) {
        // Check if type already exists
        if (typeFormationRepository.findByName(request.getName()).isPresent()) {
            throw new RuntimeException("Type de formation already exists with name: " + request.getName());
        }

        TypeFormation type = new TypeFormation();
        type.setName(request.getName());
        
        type = typeFormationRepository.save(type);
        log.info("✅ Type de formation created: {}", type.getName());
        
        return mapToResponse(type);
    }

    /**
     * Delete a type formation
     */
    @Transactional
    public void deleteType(Long id) {
        TypeFormation type = typeFormationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Type de formation not found with id: " + id));
        
        typeFormationRepository.delete(type);
        log.info("✅ Type de formation deleted: {}", type.getName());
    }

    private TypeFormationResponse mapToResponse(TypeFormation type) {
        TypeFormationResponse response = new TypeFormationResponse();
        response.setId(type.getId());
        response.setName(type.getName());
        return response;
    }
}

File: redboost_frontend/src/app/pages/backoffice/event_organizer/type-formation.service.ts
// type-formation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environment';

export interface TypeFormation {
  id: number;
  name: string;
}

export interface CreateTypeFormationRequest {
  name: string;
}

@Injectable({
  providedIn: 'root'
})
export class TypeFormationService {
  private apiUrl = `${environment.apiUrl}/type-formation`;

  constructor(private http: HttpClient) {}

  getAllTypes(): Observable<TypeFormation[]> {
    return this.http.get<TypeFormation[]>(this.apiUrl);
  }

  createType(request: CreateTypeFormationRequest): Observable<TypeFormation> {
    return this.http.post<TypeFormation>(this.apiUrl, request);
  }

  getTypeById(id: number): Observable<TypeFormation> {
    return this.http.get<TypeFormation>(`${this.apiUrl}/${id}`);
  }

  deleteType(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}